llvm: fix (and test) static libc++ and LLVM runtimes

chandlerc opened the pull request:
This adds the necessary CMake flags to pretty thoroughly enable linking
statically against libc++. With this, you can use `clang++` driver flags
to handle most aspects of the linking, and correctly get all the
relevant libraries. They also will reliably link shared objects. I've
added tests that both check the static linking can be used in position
independent code contexts and that check that none of `libc++`,
`libstdc++`, or `libgcc` end up linked when using these modes.

Details of the exact `clang++` flags needed are in the tests.

Some of the CMake flags here are really working around bugs:
https://bugs.llvm.org/show_bug.cgi?id=43604
https://bugs.llvm.org/show_bug.cgi?id=46321

For bug 43604, it seems unlikely to impact Homebrew unless the CMake
used for some reason changed its default, but it seemed worth fixing
this in place so that it works reliably.

For bug 46321, I've figured out the complex set of flags that cause
everything to work here. Hopefully the bug ends up with upstream
standardizing on this packaging approach and making the flags more
reasonable.

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [x] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

danielnachun commented:
This looks great.  Hopefully it will eventually be possible for us to also build shared libraries for the Clang runtime that don't have any dependency on GCC but you mentioned there are bugs and missing features which mean this is currently not possible on Linux (perhaps this requires a 2-stage build too?).  Providing a static `libc++` which is itself statically linked to the GCC runtime libs should at least allow users to build binaries with Clang that don't have a runtime dependency on GCC, which is rather unintuitive unless the user has a pretty solid understanding of compiler toolchains on Linux.

As with the previous PR, LLVM itself builds fine and passes all the tests.  The test failures are all unrelated problems in reverse dependences of LLVM.  This looks ready to merge to me.  @iMichka and @carlocab any thoughts?

carlocab commented on Formula/llvm.rb:
```diff
@@ -275,10 +279,59 @@ def caveats
            "-std=c++11", "-stdlib=libc++", "test.cpp", "-o", "testlibc++",
-           "-Wl,-rpath,#{opt_lib}"
+           "-L#{opt_lib}", "-Wl,-rpath,#{opt_lib}"
     on_macos { assert_includes MachO::Tools.dylibs("testlibc++"), "#{opt_lib}/libc++.1.dylib" }
```
Is it worth doing this here:
```suggestion
    assert_includes (testpath/"testlibc++").dynamically_linked_libraries, opt_lib/"libc++.1.dylib"
```

chandlerc replied:
It isn't always spelled this way, and the spelling is *technically* stable, but didn't seem worth encoding in a test.

But I could replace this with code like below that does `/#{opt_lib}/libc\+\+/` in a match in a loop if you like?

carlocab replied:
Ah, yes, whoops, that will need to be
```suggestion
    assert_includes (testpath/"testlibc++").dynamically_linked_libraries, opt_lib/shared_library("libc++", "1")
```

carlocab commented on Formula/llvm.rb:
```diff
@@ -275,10 +279,59 @@ def caveats
+             "-stdlib=libc++", "-rtlib=compiler-rt", "-unwindlib=libunwind",
+             "-static-libstdc++", "-static-libgcc", "-lpthread", "-ldl"
+      assert_equal "Hello World!", shell_output("./test_pie_runtimes").chomp
+      Pathname.new("test_pie_runtimes").dynamically_linked_libraries.each do |lib|
```
If you're applying my suggestion above, let's make this consistent:
```suggestion
      (testpath/"test_pie_runtimes").dynamically_linked_libraries.each do |lib|
```

chandlerc replied:
I'm a total Ruby novice, so any suggestions on better code are very much welcome. =D

carlocab commented on Formula/llvm.rb:
```diff
@@ -275,10 +279,59 @@ def caveats
+             "test_plugin_main.c", "-o", "test_plugin_libc++",
+             "test_plugin.so", "-Wl,-rpath=#{testpath}", "-rtlib=compiler-rt"
+      assert_equal "Hello Plugin World!", shell_output("./test_plugin_libc++").chomp
+      Pathname.new("test_plugin.so").dynamically_linked_libraries.each do |lib|
```
```suggestion
      (testpath/"test_plugin.so").dynamically_linked_libraries.each do |lib|
```
Ditto.

carlocab reviewed (commented):
LGTM, but one minor suggestion (plus two cosmetic ones).

chandlerc commented:
PTAL -- only change since the last full test run is somewhat trivial assertion improvement. The full test run failed but only because of a non-LLVM package that seems completely independent of this change.

carlocab commented on Formula/llvm.rb:
```diff
@@ -281,7 +281,7 @@ def caveats
            "-L#{opt_lib}", "-Wl,-rpath,#{opt_lib}"
     assert_includes (testpath/"testlibc++").dynamically_linked_libraries,
-                    opt_lib/shared_library("libc++", "1")
+                    (opt_lib/shared_library("libc++", "1")).to_path
```
Bit confused by this, since `opt_lib/shared_library("libc++", "1")` should already be a `Pathname`. Do you need it to be a string? (Though I would've thought it would readily be coerced into one when needed.)

chandlerc replied:
Yeah, without the `.to_path` I get an error:

```
Minitest::Assertion: Expected ["/home/chandlerc/.linuxbrew/opt/llvm/lib/libc++.so.1", "/home/chandlerc/.linuxbrew/opt/llvm/lib/libunwind.so.1", "/home/chandlerc/.linuxbrew/opt/llvm/lib/libc++abi.so.1", "/lib/x86_64-linux-gnu/libm.so.6", "/lib/x86_64-linux-gnu/libgcc_s.so.1", "/lib/x86_64-linux-gnu/libc.so.6"] to include #<Pathname:/home/chandlerc/.linuxbrew/opt/llvm/lib/libc++.so.1>.
```

(Again, I'm a Ruby novice, so I was just hitting this with the hammer indicated by my testing...)

chandlerc replied:
I guess I could map the dynamically linked libraries array into an array of `Pathname`, but that seemed like more code.

carlocab replied:
Ah. What you want there is `#to_s`:
```suggestion
                    (opt_lib/shared_library("libc++", "1")).to_s
```
which converts the `Pathname` to a string.

chandlerc replied:
Sure, the docs I read seemed to indicate that `to_path` was the canonical name and `to_s` an alias, but 🤷🏻

carlocab commented:
`castxml` test failed, but the `llvm` build and test succeeded. Seems like this might be good to go?

danielnachun commented:
Yes the `castxml` failure is because we need to add additional LDFLAGS now that we are building LLVM with GCC-11.  Not related to this PR or the previous one.  I'm merging as this looks good to go.

danielnachun commented:
@chandlerc can you do two things so we can merge this:

1) Do another version bump.

2) Squash all your commits into one commit, using `git rebase --interactive origin/master`.

chandlerc commented:
> @chandlerc can you do two things so we can merge this:
> 
> 1. Do another version bump.
> 2. Squash all your commits into one commit, using `git rebase --interactive origin/master`.

Awesome, done!

danielnachun commented:
Merged!  Thank you so much once again for all of your contributions on this and the previous PR!
